pandas-stubs/_libs/lib.pyi

```diff
@@ -1,9 +1,27 @@
+from enum import Enum
+from typing import (
+    Final,
+    Literal,
+)
+
+from pandas import Interval
+from typing_extensions import (
+    TypeAlias,
+    TypeGuard,
+)
+
+class _NoDefault(Enum):
+    no_default = ...
+
+no_default: Final = _NoDefault.no_default
+NoDefault: TypeAlias = Literal[_NoDefault.no_default]
+
 def infer_dtype(value: object, skipna: bool = ...) -> str: ...
 def is_iterator(obj: object) -> bool: ...
 def is_scalar(val: object) -> bool: ...
 def is_list_like(obj: object, allow_sets: bool = ...) -> bool: ...
-def is_interval(val: object) -> bool: ...
-def is_complex(val: object) -> bool: ...
-def is_bool(val: object) -> bool: ...
-def is_integer(val: object) -> bool: ...
-def is_float(val: object) -> bool: ...
+def is_interval(val: object) -> TypeGuard[Interval]: ...
+def is_complex(val: object) -> TypeGuard[complex]: ...
+def is_bool(val: object) -> TypeGuard[bool]: ...
+def is_integer(val: object) -> TypeGuard[int]: ...
+def is_float(val: object) -> TypeGuard[float]: ...
```

Comment:
I think most of these can also return the equivalent numpy version.

xref https://github.com/pandas-dev/pandas/pull/51309

Reply:
I made them equivalent to what is in the PR. Tried for `is_float()` using `TypeGuard[float | np.floating]`, but that fails because `decimal.Decimal(x)` fails when `x` is a numpy float.

Also merged in recent PR for `astype()`, allowing that to now be tested.

Reply:
The following seems to work just fine
```py
import numpy as np
from pandas.core.dtypes.common import is_float

is_float(np.float32(1))
```